pkg/controllers/management/gke/gke_cluster_handler.go

```diff
@@ -0,0 +1,701 @@
+			return cluster, err
+		}
+
+		cluster, err = e.setUnknown(cluster, apimgmtv3.ClusterConditionWaiting, "Waiting for API to be available")
+		if err != nil {
+			return cluster, err
+		}
+
+		gkeClusterConfigDynamic, err = buildGKECCCreateObject(cluster)
+		if err != nil {
+			return cluster, err
+		}
+
+		gkeClusterConfigDynamic, err = e.dynamicClient.Namespace(namespace.GlobalNamespace).Create(context.TODO(), gkeClusterConfigDynamic, v1.CreateOptions{})
+		if err != nil {
+			return cluster, err
+		}
+
+	}
+
+	gkeClusterConfigMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&cluster.Spec.GKEConfig)
+	if err != nil {
+		return cluster, err
+	}
+
+	// check for changes between gke spec on cluster and the gke spec on the gkeClusterConfig object
+	if !reflect.DeepEqual(gkeClusterConfigMap, gkeClusterConfigDynamic.Object["spec"]) {
+		logrus.Infof("change detected for cluster [%s], updating GKEClusterConfig", cluster.Name)
+		return e.updateGKEClusterConfig(cluster, gkeClusterConfigDynamic, gkeClusterConfigMap)
+	}
+
+	// get gke Cluster Config's phase
+	status, _ := gkeClusterConfigDynamic.Object["status"].(map[string]interface{})
+	phase, _ := status["phase"]
+	failureMessage, _ := status["failureMessage"].(string)
+	if strings.Contains(failureMessage, "403") {
+		failureMessage = fmt.Sprintf("cannot access gke, check cloud credential: %s", failureMessage)
+	}
+
+	switch phase {
+	case "creating":
+		// set provisioning to unknown
+		cluster, err = e.setUnknown(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		if err != nil {
+			return cluster, err
+		}
+
+		if cluster.Status.GKEStatus.UpstreamSpec == nil {
+			cluster, err = e.setInitialUpstreamSpec(cluster)
+			if err != nil {
+				return cluster, err
+			}
+			return cluster, nil
+		}
+
+		e.clusterEnqueueAfter(cluster.Name, enqueueTime)
+		if failureMessage == "" {
+			logrus.Infof("waiting for cluster GKE [%s] to finish creating", cluster.Name)
+			return e.setUnknown(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		}
+		logrus.Infof("waiting for cluster GKE [%s] create failure to be resolved", cluster.Name)
+		return e.setFalse(cluster, apimgmtv3.ClusterConditionProvisioned, failureMessage)
+	case "active":
+		if cluster.Status.GKEStatus.UpstreamSpec == nil {
+			cluster, err = e.setInitialUpstreamSpec(cluster)
+			if err != nil {
+				return cluster, err
+			}
+			return cluster, nil
+		}
+
+		if apimgmtv3.ClusterConditionPending.IsUnknown(cluster) {
+			cluster = cluster.DeepCopy()
+			apimgmtv3.ClusterConditionPending.True(cluster)
+			cluster, err = e.clusterClient.Update(cluster)
+			if err != nil {
+				return cluster, err
+			}
+		}
+
+		if cluster.Spec.GKEConfig.Imported {
+			if apimgmtv3.ClusterConditionPending.IsUnknown(cluster) {
+				cluster = cluster.DeepCopy()
+				apimgmtv3.ClusterConditionPending.True(cluster)
+				cluster, err = e.clusterClient.Update(cluster)
+				if err != nil {
+					return cluster, err
+				}
+			}
+		}
+
```

Comment:
Shouldnt we be making sure there is at least one nodegroup here ?

Reply:
The validator in `pkg/api/norman/customization/cluster/validator.go` ensures there is at least one node pool.